src/liballoc/tests/btree/map.rs

```diff
@@ -116,6 +136,93 @@ fn test_iter() {
     test(size, map.into_iter());
 }
 
+fn test_iter_mut<T>(size: usize)
+where
+    T: Copy + Debug + Ord + TryFrom<usize>,
+    <T as std::convert::TryFrom<usize>>::Error: std::fmt::Debug,
+{
+    let zero = T::try_from(0).unwrap();
+    let mut map: BTreeMap<T, T> = (0..size).map(|i| (T::try_from(i).unwrap(), zero)).collect();
+
+    // Forward and backward iteration sees enough
+    assert_eq!(map.iter_mut().count(), size);
+    assert_eq!(map.iter_mut().rev().count(), size);
+
+    // Iterate forwards, trying to mutate to unique values
+    for (i, (k, v)) in map.iter_mut().enumerate() {
+        assert_eq!(*k, T::try_from(i).unwrap());
+        assert_eq!(*v, zero);
+        *v = T::try_from(i + 1).unwrap();
+    }
+
+    // Iterate backwards, checking that mutations succeeded and tring to mutate again
+    for (i, (k, v)) in map.iter_mut().rev().enumerate() {
+        assert_eq!(*k, T::try_from(size - i - 1).unwrap());
+        assert_eq!(*v, T::try_from(size - i).unwrap());
+        *v = T::try_from(2 * size - i).unwrap();
+    }
+
+    // Check that backward mutations succeeded
+    for (i, (k, v)) in map.iter_mut().enumerate() {
+        assert_eq!(*k, T::try_from(i).unwrap());
+        assert_eq!(*v, T::try_from(size + i + 1).unwrap());
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord)]
+#[repr(align(32))]
+struct Align32(usize);
+
+impl TryFrom<usize> for Align32 {
+    type Error = ();
+
+    fn try_from(s: usize) -> Result<Align32, ()> {
+        Ok(Align32(s))
+    }
+}
+
+#[test]
+fn test_iter_mut_u8() {
+    test_iter_mut::<u8>(12)
+}
+
+#[test]
+fn test_iter_mut_u16() {
+    test_iter_mut::<u16>(33)
```

Comment:
Note I picked these sizes randomly to get some variation in the kind of tree depths and traversals, but I'm not at all sure there is any more coverage.

Reply:
Seems reasonable. I'd personally make them all one `#[test]` function I think, but it doesn't really matter.

Reply:
Can you also add some tests that mutably iterate over the empty `BtreeMap`? Or do such tests exist already?

Reply:
Oh boy, what a disaster... I was sure I found such a test existed while adding test_into_key_slice_with_shared_root... and that the code did ensure_root_is_owned, but none of that is true today or ever.

I also did not quite register the consequences when I noticed that `python x.py test src/liballoc --stage 0` ignores debug_assert. Basically, the rest of the world ignores them too, including those `pr` checks run here, because they light up brightly when you s/debug_assert/assert/.

Reply:
> I also did not quite register the consequences when I noticed that python x.py test src/liballoc --stage 0 ignores debug_assert. Basically, the rest of the world ignores them too, including those pr checks run here, because they light up brightly when you s/debug_assert/assert/.

Some CI builders use debug assertions, but indeed the one running here in the PR does not.
Locally you can enable debug assertions in your `config.toml`.

Reply:
Oh, okay. The offending branch was actually outside of the PR to begin with.

I guess we need to remember to turn off debug-assertions when benchmarking, because it doesn't seem to build separately, like ordinary cargo does.